src/components/Layout.re

```diff
@@ -1,4 +1,120 @@
-[@bs.module "./layout"] [@react.component]
-external make: (~children: React.element) => React.element = "default";
+open Helmet;
+
+[@bs.module "gatsby"] external useStaticQuery: string => 'a = "useStaticQuery";
+
+%bs.raw
+{|
+    import  {graphql}  from "gatsby"
+    import 'normalize.css'
+    import './layout.css'
+    import './fonts.css'
+    import './text.css'
+|};
+
+type siteMetadataType = {title: string};
+type siteType = {siteMetadata: siteMetadataType};
+type queryResponseType = {site: siteType};
+
+[@react.component]
+let make = (~children: ReasonReact.reactElement) => {
+  let data: queryResponseType =
+    useStaticQuery(
+      [%bs.raw
+        {|
+                graphql`
+                    query SiteTitleQuery {
+                      site {
+                        siteMetadata {
+                          title
+                        }
+                      }
+                    }
+                  `
+                |}
+      ],
+    );
+
+  let meta = [|
+    {
+      name: Some("description"),
+      property: None,
+      content: "A playful avatar generator for the modern age.",
+    },
```

Comment:
As we have an array of records that contains elements of type:

```ts
{
      name: string,
      content: string,
}
```
or
```ts
{
      property: string,
      content: string,
}
```

I wrote a record type with two Variant fields:

```reason
type metaItem = { 
  name: option(string),
  property: option(string),
  content: string,
};
```
And then I completed the array to match the `metaItem` type.

I am new to Reason and I am wondering if there is a better solution to fix the problem. 
I would appreciate some help on this 👍

Reply:
Hey @ayshiff,

The type that you implement works fine, but often isn't the right way of typing. One of the rules that benefit maintainability (not on this case, but in any other) is to represent with types only valid scenarios, so in this case metaItem could be `{name: Some("blaba"), property: Some("blasl"), content: "morecontent"}` which is wrong in the HTML spec.

In order to avoid that, I would recommend that you have a chance to read through Variants in Reason. But let's try to solve it in a better way:

In general speaking, variants allow defining "kinds" of the same data structure, in this case:

```
type metaItem = | NameProperty | PropertyContent;
```
I just made up the names, but probably there are better names for this.

Since Variants can contain other data structures, we could store the name/property/content inside each variant:
```
type metaItem = | NameProperty(string, string) | PropertyContent(string, string);
							/* name, property */           /* property, content */
```

and the data that you created looks like:

```[NemaProperty("", ""), PropertyContent("", "")]```


With this structure, we might have an issue since our types don't match with the actual objects from react-helmet, now it's just time to create a "translater" function.

```
let metaToHelmet = (item: metaItem): {"name", "property", "content"}
```

---

With all this long explanation, I would say there's a tradeoff here that all the looping through all the metaItems, so remember to keep this list outside of the component just to keep things clean.

Hope it helped

Reply:
Thanks for your great explanations @davesnx 👍

I did something like this: 

```reason
type metaItem =
  | MetaWithName(string, string)
  | MetaWithProperty(string, string);

type meta = array(metaItem);

let meta: meta = [|
  MetaWithName("twitter:card", "summary"),
  MetaWithName("twitter:site", "@draftbit"),
  MetaWithName("twitter:title", "Personas by Draftbit"),
  // ...
  MetaWithProperty("fb:app_id", "292869531545861"),
  MetaWithProperty("og:type", "website"),
  // ...
|];

type finalMetaItem = {
  name: option(string),
  property: option(string),
  content: string,
};

let newMeta =
  meta
  |> Array.map(el => {
       switch (el) {
       | MetaWithName(name, content) => {
           name: Some(name),
           property: None,
           content,
         }
       | MetaWithProperty(property, content) => {
           property: Some(property),
           name: None,
           content,
         }
       }
     });
```

But again I need to do something with the Variant behavior of certain fields.

As I understand from your `metaToHelmet` "translater", it always returns something with the shape `{"name", "property", "content"}` but since we have cases where the values must not be present in the record, I don't understand how we can handle the presence or not of the value other than with Variants.

[Playground reproduction](https://rescript-lang.org/try?ext=re&code=C4TwDgpgBAthwEMCSwIygXgFBSgHygFl4EB1AS2AAsA5BOACgGdgAncgOwHMAaKF9twCUOfERIVqABVYB7SK1DM2nXvxXCA3FiyhIsEpigJWrBCAZxEKNEO1YANvAOIAXC4RGA2nlHFEkrT0EAyiuABEACYQTADG7GDA5LIc4TxhUOEAglBgDuYAZgCuDsYAbgiIrFBcEBwQZsCy1QXNUNTQMLLRrBzGtQB0aaJC6bj+ZJRBjOHAAO6UqKyusSaRaZlMRTAwJiDho34SU3Qz84sNrkyUEBvhAAKRZgXAAEaUB2PiASfBDLMLYBLVxJYBOO5SBpMFIIJhQV4gKAAEWebw+h3Gx2opxCGQBF2W5F2tWGuAiVCBYCYrgA9DSFNCOLCBk8EC93sABrFZDAaXMIK8ALRgVgQMrkCBzQUARgAHtKBmBuKSoBjvpNsX88ecgZdonEEkkUiqIjk8oUSuVKiYanUGpU2q1qh1YN0Gn0EIMVWqJoEcaEyZkANYQEBzZqRJgmzJZCpVPix63VADidsazT4kNYjNhfBRbLRwATcZMUa+PqxVBk8gaSnCBVergQYDAAH1yOs+OEAEwATm7AA4AGy9gCsAGZpaOACyj4fSz5HH7SOQKOuyLgg8C3Lv817XVCLzHLqur2sWPEb1xEz07vEU4BU2n0qEwpgs1Ecrk8vkC4Wi8VJUVZVyy+X0pmrNcLHCK9QXBLssxzOEEWRT90TAytIPPf4ryKVgHDuB8nzpBk3w-Asv25GAj3VQIsMUC9Axgzd9XichEmSVIvlNXJ8hAYpSgQEtqlqep0xaNoXS6HoPS9UCsDwABdew9GgApOAQBwJhsdAMCgABvUQmTgdx5CNDhlEELg1RFGsGNMjiUks1Q1W5DhUHc9wBFUdIAF97CcYAoHqOYJkwUQrAQUQ8AAPigLJTHMAZdjABgIFKDA4sMwMySYQFYioKA0ocIQDIyMkCHAzVGGMiA+DcjzgFKzKypytravcABlHkQlqtU2rJWyoPcGgUjq8q2oauoiwmslfNmyrMLPBiGCG896pSRrmqy2bAzW+yoG6xh9tAfqBuC4IRrG7jzqm9ybpy+aBqewNfLsLAgA)

Reply:
Exactly, using records for the "finalMetaItem" doesn't fit the scenario. I would use bs.deriving abstract for now, which transforms the record into a module with a method that creates those optionals for you. I can't find any good quality documentation now, but I'm sure you could find more info in the rescript docs (even if we are using bs v7).

https://rescript-lang.org/try?ext=re&code=C4TwDgpgBAthwEMCSwIygXgFBSgHygFl4EB1AS2AAsA5BOACgGdgAncgOwHMAaKF9twCUOfERIVqABVYB7SK1DM2nXvxXCA3FiyhIsEpigJWrBCAZxEKNEO1YANvAOIjAbTyjiiSbXoQGUVwAIgATCCYAY3YwYHJZDmCeIKhggEEoMAdzADMAVwdjADcERFYoLggOCDNgWXKc+qhqaBhZcNYOY0qAOiTRIWTcbzJKP0Zg4AB3SlRWAC5Ik1Ck1KY8mBgTEGDBrwkxugnp2Zr5pkoIVeCAAVCzHOAAI0pdofEfQ-8GSZngOfmcWATmuUhqTASCCYUCeICgABEHs9XnthgdqEcAilfqcFuQtpV+rgQlR-mAmPMAPSUhQQjhQnr3BCPF7AHqRWQwSlTCBPAC0YFYECK5AgUz5AEYAB4SnpgbhEqCoj6jDHfbEnf5ncJRGJxBKKkIZLK5ArFUomCpVGqlJqNcotWDtGpdBC9RXKka+TGBYmpADWEBAU3qoSYhtSaRKZT4UYt5QA4tbavU+GDWHSoXxEczkcBY9GTOH3p70VQZPIakpgjknvMEGAwAB9cgrPjBABMAE4OwAOABsXYArABmCVDgAsQ4HEre+0+0jkCmrsi4gPAV3bPKeF1Qc7RC-LS6rFmxq-m+Ldm+xpOA5KpNPBkKYjKRrPZnO5vIFQpFYrlColu8XpjBWy4WME55AiC7bppm0Kwgib4osBZZgSePznnkrAONct73tStLPq+ubvhyMD7iqvjoYop5+pBa46tE5CxPEiTvEamTZCA+SFAghblJU1Qpg0TSOm0HSuu6QFYHgAC69huDcO6MjU5Aitwxg7mwCCRMAcm6BuUA5JwCAOCMNjoBgUAAN6iEpKnyPq9IOAZuD0nA8zqIIvD2cpL5OWxZluZkx60V5AiqO8DkBaxkKuaIHIcKgyURRovkAL72E4wBQNUUwjJgohWAgoh4AAfFAaSmOYPRbGADAQIUGCVXZfrEkwfyRFQUCNQ4Qi2SkxIECBaqMB5EB8ElKXAANLWDe1i0mS5FmoDADAAH4TXwG3TVU+a9UIdhDX6GUnWIo1HpWtEMIK12gFNCQzXNrXnX6y1mataCbXd4E7XtyV8AwR3aItxJnWDEOncdWAAFIvg4q4MPlIzHUAA